### PR TITLE
Add funcationality for DRE to python API

### PR DESCRIPTION
**Pull request by abachma2**
This PR adds some functionality for the DRE to the python API that was missing. This includes:
- `AddMutualReqs` added to the `RequestPortfolio` class -- was previously defined but not used anywhere
- `AddRecipe` added to 'lib' class (`cyclus/lib/pyx`)
- Add the ability to specify a request as exclusive.

To implement the `AddMutualReqs` and the exclusivity of the request, the structure of the dictionary to define a request, was modified to properly attach a material preference to each commodity that can meet a request and define which commodity requests are mutual.  This dictionary is read into the `request_request_portfolio_to_cpp` and `material_request_portfolio_to_cpp` functions in the `cyclus/typesystem.py` file.  

The previous format for this dictionary was: `[{"commodities":{'commod_name':Material object,...},"constraints":number},...]`. The new format is: `[{"commodities":[{'commod_name':Material object,"preference":number, "exclusive":Bool},...], "constraints":number}]`. 

In the previous format, each outer dictionary (list element) was treated as a separate request and each commodity and Material pair in the nested dictionary was treated with AND logic, such that each material was part of the entire request to be made. In the new format, each outer dictionary is still treated as separate requests. The nested inner list of dictionaries assigns the preference of each commodity, and define which commodity requests should be mutual requests (apply an OR logic to them). This new format helps to keep preferences and exclusive settings paired with each request.

The `AddRecipe` functionality is less related to the DRE, but add the ability to record recipes to the output database during a 
simulation. This was already defined in the c++ API.

### CODE REVIEW

**Comment by gonuke** [comment]
If you rebase this on the updates - we can look to see where this stands?

**Comment by abachma2** [comment]
Yeah. I'll take care of this next week once I take care of some other stuff. Feel free to ping me if you don't see any movement on this by ext Wednesday (10/4).

**Comment by abachma2** [comment]
I've made this PR open for review. The build and test CI isn't passing and I think it's because it's having issues building the docker images. I'll look into this more this afternoon.

File: README.rst [hunk 1]
```diff
@@ -79,7 +79,7 @@ Package                Minimum Version
 ``boost``                1.46.1
 ``libxml2``              2
 ``libxml++``             2.36
-``python``               2.7 or 3.3+
+``python``               2.7 or 3.10
```
Comment (gonuke) [inline README.rst:82]: Haven't we moved on from Python 2?

Maybe I need to update this for #1588

File: CHANGELOG.rst [hunk 2]
```diff
@@ -17,6 +17,7 @@ Since last release
   will fail unless update-alternatives has been used to point python at the 
   correct python3 version (#1558)
 * build and test are now fown on githubAction in place or CircleCI (#1569)
+* AddMutalReqs and AddReciepe functions and exclusive bids in python API of DRE (#1584)
```
Comment (gonuke) [inline CHANGELOG.rst:20]:
```suggestion
* AddMutalReqs and AddRecipe functions and exclusive bids in python API of DRE (#1584)
```

**Review by gonuke — CHANGES_REQUESTED** [review]
Thanks for digging in to this esoteric and deep stuff @abachma2.  Can you point to where this Python data structure is now matching something that exists already in C++?

**Comment by gonuke** [comment]
> I've made this PR open for review. The build and test CI isn't passing and I think it's because it's having issues building the docker images. I'll look into this more this afternoon. 

Hmmm... I think this is a flaw in my CI plan!  Something I should have know from earlier experiences.  We may need to separate the docker building from the CI again, after all.

You might be able to make progress on this by pushing this as a branch to the upstream and then making a PR from that branch.

File: README.rst (hunk 1 above)
Comment (abachma2) [reply README.rst:82]: I don't know. I've only ever used cyclus with python 3. If we know it's no longer supported then I'll remove it.

**Comment by abachma2** [comment]
The contents of this branch have been moved to a remote branch, to allow for any potential changes to the Docker image for CI. Therefore, I am closing this PR. Please refer to #1594 for continued discussion and changes.

**Pull request closed without merging**